packages/react-components/src/components/xy-output-component.tsx

```diff
@@ -273,21 +283,123 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
         this.setState({orderedNodes: ids});
     }
 
-    private beginSelection(event: React.MouseEvent<HTMLDivElement, MouseEvent>) {
+    private onMouseDown(event: React.MouseEvent<HTMLDivElement, MouseEvent>) {
         this.mouseIsDown = true;
         this.posPixelSelect = event.nativeEvent.screenX;
-        const offset = this.props.viewRange.getOffset() ?? 0;
-        const scale = this.props.viewRange.getEnd() - this.props.viewRange.getstart();
-        const xPos = this.props.viewRange.getstart() - offset +
-            (event.nativeEvent.offsetX / this.lineChartRef.current.chartInstance.width) * scale;
+        const startTime = this.getTimeX(event.nativeEvent.offsetX);
         this.props.unitController.selectionRange = {
-            start: xPos,
-            end: xPos
+            start: startTime,
+            end: startTime
         };
         document.addEventListener('mousemove', this.updateSelection);
         document.addEventListener('mouseup', this.endSelection);
     }
 
+    private updateRange(rangeStart: number, rangeEnd: number) {
+        if (rangeEnd < rangeStart) {
+            const temp = rangeStart;
+            rangeStart = rangeEnd;
+            rangeEnd = temp;
+        }
+        this.props.unitController.viewRange = {
+            start: rangeStart,
+            end: rangeEnd
+        };
+    }
+
+    private zoom(iszoomIn: boolean) {
+        if (this.props.unitController.viewRangeLength >= 1) {
+            let newStartRange = 0;
+            let newEndRange = 0;
+            let zoomNumber = 1;
+            const position = this.getTimeX(this.positionXMove);
+            if (!iszoomIn) {
+                zoomNumber = -1;
+            }
+            const startPos = position - this.props.unitController.viewRange.start;
+            const endPos = this.props.unitController.viewRange.end - position;
+            newStartRange = position - (startPos - zoomNumber * (startPos * TEN_PERCENT));
+            newEndRange = position + (endPos - zoomNumber * (endPos * TEN_PERCENT));
+            this.updateRange(newStartRange, newEndRange);
```

Comment:
This creates a strange behavior in the following use case:
- Zoom to half the trace range
- Pan to beginning of trace
- Try to zoom out with 'S" while the mouse is very close to the right edge.

What happens is that the window range start is beyond the start time (< 0) and gets snapped back to 0 in time-graph-unit-controller. Then the effect is a very slow zoom out step.

What I would suggest is to calculate the new view range length based on zoom in/out (capped to absolute range), then apply this new window range centered on the mouse position. After this, snap the new window range to the bounds [0, absolute range] like it's done below for the pan case.